IMDS v2

hankjacobs opened the pull request:
Add support for enforcing tokens on the metadata service per [v2 of the IMDS API](https://aws.amazon.com/blogs/security/defense-in-depth-open-firewalls-reverse-proxies-ssrf-vulnerabilities-ec2-instance-metadata-service/)

hankjacobs commented:
Mostly done. Just need to add VCR tests.

hankjacobs commented:
The downside with this approach is the token validity is limited to the duration of which the metatron cert is valid and kept. I believe it’s currently only kept for 5 minutes. I may switch implementations to just keeping track of random tokens.

hankjacobs commented:
I also realized I still need to add support for fetching tokens when proxying requests

hankjacobs commented on metadataserver/server.go:
```diff
@@ -25,6 +25,7 @@ import (
 const defaultMacAddress = "00:00:00:00:00:00"
 
 var whitelist = set.NewSetFromSlice([]interface{}{
+	"/latest/api/token",
```
This should be removed. This should never be proxied

hankjacobs commented:
Better yet, I’m going to switch to a unique key generated at startup + HMAC

sargun commented on .golangci.yml:
```diff
@@ -16,3 +16,6 @@ run:
     - vpc/limitsold
+issues:
+  exclude:
+    - Using the variable on range scope .* in function literal
```
This has caught a number of real issue (primarily around launching goroutines in functions). I'd rather not exclude this.

sargun commented on cmd/titus-metadata-service/main.go:
```diff
@@ -213,6 +214,11 @@ func main() {
+		cli.BoolFlag{
+			Name:        "require-token",
+			Usage:       "Set to true to require a token",
+			Destination: &requireToken,
```
Adding an EnvVar here would be nice.

sargun commented on metadataserver/auth/auth.go:
```diff
@@ -0,0 +1,165 @@
+	mac := a.hmac(envEnc)
+	sig := base64.StdEncoding.EncodeToString(mac)
+
+	return envEnc + "." + sig, nil
```
It looks like you've reimplemented / reinvented JWT / JWS here. Any reason not to just use off the shelf JWTs? They have expiration and stuff "built-in" to them?

sargun replied:
(I used the first implementation of Go JWT examples I found)
https://play.golang.com/p/NhQDPykRQdY


Although your (our) certificate signer implemens a slightly different interface from the one they expect. I imagine we can map it across.

hankjacobs replied:
Yep, will use JWT

sargun commented on metadataserver/authentication.go:
```diff
@@ -0,0 +1,78 @@
+
+		auth := auth.HMACAuthenticator{Key: ms.tokenKey}
+		if !auth.VerifyToken(token) {
+			http.Error(w, "", http.StatusUnauthorized)
```
It might be worth indicating if the token is totally invalid vs. not present vs. expired

hankjacobs replied:
This is the way IMDS v2 actually behaves. I’m concern we’d break clients if we deviate.

sargun replied:
Can we log the reason at least?

sargun replied:
All the logs end up in journald, and we can setup alerting based on it.

hankjacobs replied:
Yep Yep, good cal. I'll do a logging pass through all of the new code.

hankjacobs replied:
Yep yep, good call. I'll do a logging pass on all of my new code. I'll take a look at adding handy metrics as well.

sargun commented on metadataserver/authentication.go:
```diff
@@ -0,0 +1,78 @@
+	}
+	defer conn.Close()
+
+	if conn.RemoteAddr().(*net.TCPAddr).IP.To4() != nil {
```
We need this to work on IPv6 (in the future as well)

hankjacobs replied:
Shouldn’t be a problem but I don’t think IMDS is available over v6. I could be wrong though.

hankjacobs commented:
Well, at the beginning I wanted to avoid them because they felt too heavy but yeah, I agree, it’s already practically a JWT. Will swap it in.

sargun commented:
> I also realized I still need to add support for fetching tokens when proxying requests

Do you mean that the system "consume" IMDSv2 on the backend? I don't think that we need to do that.

sargun commented:
It seems like it's in the middle of moving to JWTs. Let me know when I should take another look.

hankjacobs commented:
JWT support is complete. The only things left are your code review changes
and updating the VCR tests. The review can totally wait until then.

On Mon, Mar 16, 2020 at 12:14 AM Sargun Dhillon <notifications@github.com>
wrote:

> It seems like it's in the middle of moving to JWTs. Let me know when I
> should take another look.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Netflix/titus-executor/pull/359#issuecomment-599379886>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFNHRE3INMYTWMQEX4M2EDRHXGT5ANCNFSM4LKRQUWQ>
> .
>

hankjacobs commented:
What's probably a little confusing is I left the certificate implementation in there. I'll gut it.

hankjacobs commented:
I've already added support for using IMDS v2 on the backend (in the proxy). It'll allow us to turn of v1 at some point in the future (I thought that was the intention and is why I implemented it).

sargun commented on .golangci.yml:
```diff
@@ -16,3 +16,4 @@ run:
     - vpc/bpf/filter
     - vpc/bpf2
     - vpc/limitsold
+    
```
Whitespace weirdness.

sargun commented on metadataserver/auth/auth.go:
```diff
@@ -0,0 +1,132 @@
+// GenerateToken token
+func (a *JWTAuthenticator) GenerateToken(ttl time.Duration) (string, error) {
+	claims := &jwt.StandardClaims{
+		ExpiresAt: time.Now().Add(ttl).Unix(),
```
Please add the task ID to the token, or similar. I think they call this "Audience" in JWT terms. I believe partially it'll improve security, but more so, improve debugging.

sargun commented on metadataserver/auth/auth.go:
```diff
@@ -0,0 +1,132 @@
+}
+
+// GenerateToken generates a token
+func (a *CertificateAuthenticator) GenerateToken(ttl time.Duration) (string, error) {
```
Am I missing something here, or isn't most of this functionality in the `jwt-go` library?

sargun replied:
Or is this being ripped out?

hankjacobs replied:
It’s being ripped out.

sargun replied:
:D

sargun commented on vendor/modules.txt:
```diff
@@ -161,18 +163,9 @@ github.com/golang/glog
 # github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e
 github.com/golang/groupcache/lru
 # github.com/golang/protobuf v1.3.3
-github.com/golang/protobuf/proto
```
Why did all these get removed?

sargun commented:
There are three make targets you might find helpful:

`lint - Runs lint on the files. Returns linting errors.`
`fmt - fmts all the files`
`protogen - generate new golang wrappers for the protobufs. Run the fmt target after this.`

hankjacobs commented:
Re #1: My thought when implementing it was that it's OK to generate a new key on startup because worse comes to worse, clients will fetch a new token. They already have to deal with token expiration and the error looks the same to the client in both cases . That being said, I don't have strong feelings about it so I'm game to do your suggestion if you do feel strongly about it.

Re #2: Great idea. It should simply the code a good deal. I'll go that route.

hankjacobs commented:
We could also use the task ID as the key that way it’s unique per task but is the same across restarts.

sargun commented:
I think we should have some kind of configurable value appended to task ID as the key for the HMAC, and then we're good to go.

hankjacobs commented:
So after going down a pretty deep rabbit hole I came to the conclusion that the aws sdk's metadata client isn't suitable for what we want to do. I could make it work but I'd have to abuse the hell out of the API it exposes. That being said, I replaced the proxying server implementation with one using an http client. I still have to fix up the tests and finish applying your feedback but its looking good so far.

sargun commented:
Looks good so far. Part of me almost thinks we should split it into two parts. 
1) Serving up the IMDSv2 API to our clients
2) Consuming the IMDSv2 API from AWS

#1 I see as almost done, except a little bit of logging, and making the key stable across restarts.

Do you want to to do that work first? and commit it, and then we can figure out the 2nd part later? I think the only thing that would be required is that the current http proxy would have to strip the token when proxying requests.

hankjacobs commented:
Sure. I'll go that route.

hankjacobs commented:
This should be good to go now.

sargun commented on cmd/titus-metadata-service/main.go:
```diff
@@ -236,6 +251,8 @@ func main() {
 			Optimistic:          optimistic,
 			APIProtectEnabled:   apiProtectEnabled,
 			StateDir:            stateDir,
+			RequireToken:        requireToken,
```
Can we add a (logging) warning is the salt is empty?

hankjacobs commented:
Actually, I want to clean up the Verify api real quick, no need for the bool

sargun commented on metadataserver/auth/auth.go:
```diff
@@ -0,0 +1,55 @@
+	})
+
+	if err != nil {
+		return false, 0, errors.Wrap(err, "Token invalid")
```
Any reason not to make the function signature (int64, error). If err != nil, then you can assume token is invalid.

sargun commented on metadataserver/auth/auth.go:
```diff
@@ -0,0 +1,55 @@
+		return false, 0, errors.Wrap(err, "Token invalid")
+	}
+
+	if !claims.VerifyAudience(a.Audience, false) {
```
Maybe errors.wrap should be errors.Wrapf("Token invalid, expected audience %q but have audience %q", claims.Audience)

sargun commented on metadataserver/auth/auth.go:
```diff
@@ -0,0 +1,55 @@
+	}
+
+	if !jwtToken.Valid {
+		return false, 0, errors.Wrap(err, "Token invalid")
```
I don't think you want to return a wrapped error here, because err will = nil based on above.

sargun commented on metadataserver/auth/auth.go:
```diff
@@ -0,0 +1,55 @@
+	}
+
+	if claims.Valid() != nil {
+		return false, 0, errors.Wrap(err, "Token invalid")
```
I think you want to return the error here that was returned from claims.Valid(), and not the one from above

sargun commented on metadataserver/authentication.go:
```diff
@@ -0,0 +1,109 @@
+		valid, remaining, err := auth.VerifyToken(token)
+		if !valid {
+			if err != nil {
+				log.Error("Token invalid: ", err)
```
Please use log.WithError(err).Error("Token invalid")

sargun commented on metadataserver/authentication.go:
```diff
@@ -0,0 +1,109 @@
+
+	forwarded := r.Header.Get("x-forwarded-for")
+	if len(forwarded) > 0 {
+		log.Error("`x-forwarded-for` header present, blocking request`")
```
please make log the value of x-forwarded-for, so something like:
`log.WithField("xForwardedFor", forwarded).Error(..)`

Also, I don't think we need the backticks.

hankjacobs replied:
The back ticks are just for stylization purposes when printed but I'll remove.

hankjacobs commented:
All cleaned up.

sargun commented on metadataserver/authentication.go:
```diff
@@ -0,0 +1,109 @@
+		return
+	}
+
+	if ttlSec < 0 || ttlSec > 21600 {
```
(6 * time.Hour).Seconds()?

sargun commented on metadataserver/authentication.go:
```diff
@@ -0,0 +1,109 @@
+	}
+
+	if ttlSec < 0 || ttlSec > 21600 {
+		log.Error("Invalid ttl: ", ttlSec)
```
Please use withfield

sargun commented on metadataserver/authentication.go:
```diff
@@ -0,0 +1,109 @@
+
+	token, err := auth.GenerateToken(time.Duration(ttlSec) * time.Second)
+	if err != nil {
+		log.Error("Could not generate token: ", err)
```
Please use witherror.

sargun commented on metadataserver/authentication.go:
```diff
@@ -0,0 +1,109 @@
+	if !ok {
+		// Not a Hijacker, just treat it as a normal ResponseWriter
+		if _, err := fmt.Fprint(w, token); err != nil {
+			log.Error("Unable to write token: ", err)
```
Please use witherror.

sargun commented on metadataserver/authentication.go:
```diff
@@ -0,0 +1,109 @@
+
+	conn, bufrw, err := hj.Hijack()
+	if err != nil {
+		log.Error("Unable to hijack connection: ", err)
```
Please use witherror.

sargun commented on metadataserver/authentication.go:
```diff
@@ -0,0 +1,109 @@
+		return
+	}
+
+	conn, bufrw, err := hj.Hijack()
```
TIL. about this API. Neat.

sargun commented on metadataserver/authentication.go:
```diff
@@ -0,0 +1,109 @@
+
+	_, err = bufrw.Write([]byte(token))
+	if err != nil {
+		log.Error("Unable to write token: ", err)
```
Please use withErr

sargun commented on metadataserver/authentication.go:
```diff
@@ -0,0 +1,109 @@
+		_ = p.SetTTL(1)
+	}
+
+	if conn.RemoteAddr().(*net.TCPAddr).IP.To16() != nil {
```
maybe else if?

sargun commented on metadataserver/authentication.go:
```diff
@@ -0,0 +1,109 @@
+	if conn.RemoteAddr().(*net.TCPAddr).IP.To16() != nil {
+		p := ipv6.NewConn(conn)
+		_ = p.SetHopLimit(1)
+	}
```
Maybe else {
log.warning("could not determine protocol type?
}

sargun commented on metadataserver/identity/verify.go:
```diff
@@ -0,0 +1,126 @@
+package identity
```
I'm a little bit confused. Is this code only for testing? Or is it used elsewhere?

hankjacobs replied:
It was used. It is no longer used. I was on the fence on whether I should remove it or not because I can see having signature verification being potentially handy at some point. I don't have a strong feeling either way. Your call.

sargun replied:
They seem useful. Maybe just add a comment indicating they're helper functions or something? I can imagine the metatron client consuming these at some point.

sargun commented:
Whoops, I reviewed a bunch of outdated bits.

sargun commented on metadataserver/server_test.go:
```diff
@@ -633,3 +635,69 @@ func TestTaskIdentityWithECDSA(t *testing.T) {
+	ms := setupMetadataServer(t, ess, ecdsaCerts[0], true)
+
+	// Get Token
+	tokenPath := fmt.Sprintf("http://%s%s", ess.proxyListener.Addr().String(), "/latest/api/token")
```
Any reason we don't use the "official" ec2metadata client here? and just override the EC2 metdata URL?

hankjacobs replied:
Good point. I’ll probably add that as another test since you can’t really inspect the response status codes directly.

hankjacobs commented:
I'm really glad you suggested using the ec2metadata client. Once a connection is hijacked, the full lifecycle of the connection and http response is the responsibility of the hijacker. My first implementation just wrote the token and none of the HTTP status or headers. That resulted in a malformed response which, funnily enough, curl just happily accepts so I didn't catch it when I was doing testing on an instance. The ec2metadata client did.

hankjacobs commented:
@sargun I'd like your eyes on the new changes to createAuthTokenHandler when you have a moment. Specifically with the manually writing the http status and headers to the hijacked connection. Thanks.

sargun commented:
Dumb question. Is the entire point of the hijacker to be able to set the connection TCP max TTL to 1?

hankjacobs commented:
Yep exactly. I almost ditched it but it does feel like a good security feature.

sargun commented:
1) Can we make the X-Forwarded-For check a warning right now, and then put it behind a command line option. We don't know if anyone in the fleet is relying on this behaviour right now, and breaking people's IMDSs sucks (it's hard for them to debug)

sargun commented:
2) You may have already looked at this option, but can we instead wrap the listener, like:
https://repl.it/repls/PaleCrushingSubweb

sargun commented:
This way, the TTL-setting happens at the listener. If you want to get really exciting, you can figure out how to set the TTL on the listener itself. (Keep in mind, this code only runs on Linux and Mac OS X).

sargun commented:
That way there's significantly less exciting issues that can occur with the world of HTTP.

sargun commented:
Right now, it's kind of jank, because the listener "work" happens in `cmd/titus-vpc-tool`, I don't see a strong reason not to pass down the "raw" (unwrapped) listener down through mdsCfg

sargun commented:
(I might be suggesting something stupid, or you may have already gone down this route. In which case, I'll look at the thing more closely. At first glance, it looks okay)
